onnx/defs/schema.cc

```diff
@@ -1535,11 +1536,51 @@ void OpSchema::Finalize() {
   ParseAndSetTypes(&inputs_);
   ParseAndSetTypes(&outputs_);
 
+  if (HasContextDependentFunction()) {
+    ENFORCE(node_determinism_ != NodeDeterminism::Unknown);
+  }
+
   for (auto& func : opset_version_to_function_body_) {
     BuildFunction(*func.second);
   }
 }
 
+OpSchema::NodeDeterminism OpSchema::GetNodeDeterminism() const {
+  if (node_determinism_ == NodeDeterminism::Unknown) {
+    if (HasContextDependentFunction()) {
+      return NodeDeterminism::Unknown;
+    } else if (const FunctionProto* func_proto = GetFunction(); func_proto) {
+      const OpSchemaRegistry& reg = *OpSchemaRegistry::Instance();
+      std::unordered_map<std::string, int> domain_to_opset_version;
+      for (const auto& opset : func_proto->opset_import()) {
+        domain_to_opset_version[opset.domain()] = opset.version();
+      }
+      for (const NodeProto& n : func_proto->node()) {
```

Comment:
TODO: This loop maybe slow in large function and after determining unknown state should be changed to determined value